Cython/Utility/ObjectHandling.c

```diff
@@ -2553,3 +2553,132 @@ static PyObject *__Pyx_PyMethod_New(PyObject *func, PyObject *self, CYTHON_UNUSE
+#else
+    #define __PYX_PYUNICODE_CONCATINPLACE __Pyx_PyUnicode_Concat
+#endif
+#define __PYX_PYUNICODE_CONCATINPLACESafe(a, b) ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
+    PyNumber_Add(a, b) : __PYX_PYUNICODE_CONCATINPLACE(a, b))
+
+////////////// UnicodeConcatInPlace ////////////////////
+
+#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION > 2
+    static CYTHON_INLINE PyObject *__Pyx_PyUnicode_ConcatInPlaceImpl(PyObject **a, PyObject *b, void * __pyx_refnanny) {
+    __Pyx_GIVEREF(*a);
+    PyUnicode_Append(a, b);  // this does it's own safety checks so is always safe to use
+    __Pyx_XINCREF(*a);  // there are two copies of `*a` - the one just returned and the one
+                // passed in as an argument
+    return *a;
+}
+#endif
+
+//////////// BytesConcat.proto ///////////////////////
+
+// Follows the interface of "PyUnicode_Concat", but with Bytes, rather than the
+// interface of PyBytes_Concat. However, uses PyBytes_Concat which has the potential
+// to modify in-place. See UnicodeConcatInPlace for comments
+#if CYTHON_COMPILING_IN_CPYTHON
+    #if CYTHON_REFNANNY
+        #define __PYX_PYBYTES_CONCATINPLACE(a, b) __Pyx_PyBytes_ConcatInPlaceImpl(&a, b, __pyx_refnanny)
+    #else
+        #define __PYX_PYBYTES_CONCATINPLACE(a, b) __Pyx_PyBytes_ConcatInPlaceImpl(&a, b, NULL)
+    #endif
+    static CYTHON_INLINE PyObject *__Pyx_PyBytes_ConcatInPlaceImpl(PyObject **a, PyObject *b, void * __pyx_refnanny); /* proto */
+    static CYTHON_INLINE PyObject *__Pyx_PyBytes_Concat(PyObject *a, PyObject *b);
+#else
+    #define __Pyx_PyBytes_Concat(a, b) PyNumber_Add(a,b)
+    #define __PYX_PYBYTES_CONCATINPLACE(a, b) PyNumber_Add(a,b)
+#endif
+#define __Pyx_PyBytes_ConcatSafe(a, b) ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
+    PyNumber_Add(a, b) : __Pyx_PyBytes_Concat(a, b))
+#define __PYX_PYBYTES_CONCATINPLACESafe(a, b) ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
+    PyNumber_Add(a, b) : __PYX_PYBYTES_CONCATINPLACE(a, b))
+
+//////////// BytesConcat ///////////////////////
+
+#if CYTHON_COMPILING_IN_CPYTHON
+static CYTHON_INLINE PyObject *__Pyx_PyBytes_ConcatInPlaceImpl(PyObject **a, PyObject *b, void * __pyx_refnanny) {
+   __Pyx_GIVEREF(*a);
+    PyBytes_Concat(a, b);  // this does it's own safety checks so is always safe to use
+    __Pyx_XINCREF(*a);  // there are two copies of `*a` - the one just returned and the one
+                // passed in as an argument
+    return *a;
+}
+
+static CYTHON_INLINE PyObject *__Pyx_PyBytes_Concat(PyObject *a, PyObject *b) {
+    // unusual for utilitycode to use refnanny, but the __Pyx_PyBytes_ConcatInPlaceImpl
+    // needs something forwarded to it
+    __Pyx_RefNannyDeclarations
+    __Pyx_RefNannySetupContext("__Pyx_PyBytes_Concat (internal)", 0);
+    // wrapper function prevents "outside" a from getting modified
+    PyObject* result;
+    Py_INCREF(a); // an extra reference will prevent PyBytes_Concat from operating in place
```

Comment:
Then why, exactly, do we do this at all?

Reply:
It does give a direct call rather than going through the operator lookup in `PyNumber_Add`, but it's probably more trouble than it's worth. I'll take it out.